File: packages/cli/internal/linker/linker.go
package linker

import (
	"fmt"
	"strings"

	"github.com/agentx-labs/agentx/internal/integrations"
	"github.com/agentx-labs/agentx/internal/userdata"
)

// typeSection maps a type reference path prefix to the ActiveConfig section name.
func typeSection(typeRef string) (string, error) {
	parts := strings.SplitN(typeRef, "/", 2)
	if len(parts) < 2 {
		return "", fmt.Errorf("invalid type reference %q: expected <type>/<name> (e.g., personas/senior-java-dev)", typeRef)
	}

	switch parts[0] {
	case "personas":
		return "personas", nil
	case "context":
		return "context", nil
	case "skills":
		return "skills", nil
	case "workflows":
		return "workflows", nil
	case "prompts":
		return "prompts", nil
	default:
		return "", fmt.Errorf("unknown type prefix %q in reference %q", parts[0], typeRef)
	}
}

// getSection returns a pointer to the slice in ActiveConfig for the given section name.
func getSection(active *ActiveConfig, section string) *[]string {
	switch section {
	case "personas":
		return &active.Personas
	case "context":
		return &active.Context
	case "skills":
		return &active.Skills
	case "workflows":
		return &active.Workflows
	case "prompts":
		return &active.Prompts
	default:
		return nil
	}
}

// contains checks if a string slice contains a value.
func contains(slice []string, val string) bool {
	for _, s := range slice {
		if s == val {
			return true
		}
	}
	return false
}

// remove returns a new slice with the given value removed.
func remove(slice []string, val string) []string {
	result := make([]string, 0, len(slice))
	for _, s := range slice {
		if s != val {
			result = append(result, s)
		}
	}
	return result
}

// AddType adds a type reference to the correct section of project.yaml and runs sync.
func AddType(projectPath, typeRef string) error {
	section, err := typeSection(typeRef)
	if err != nil {
		return err
	}

	config, err := LoadProject(projectPath)
	if err != nil {
		return err
	}

	target := getSection(&config.Active, section)
	if target == nil {
		return fmt.Errorf("internal error: unknown section %q", section)
	}

	if contains(*target, typeRef) {
		return fmt.Errorf("%s is already linked", typeRef)
	}

	*target = append(*target, typeRef)

	if err := SaveProject(projectPath, config); err != nil {
		return err
	}

	return Sync(projectPath)
}

// RemoveType removes a type reference from project.yaml and runs sync.
func RemoveType(projectPath, typeRef string) error {
	section, err := typeSection(typeRef)
	if err != nil {
		return err
	}

	config, err := LoadProject(projectPath)
	if err != nil {
		return err
	}

	target := getSection(&config.Active, section)
	if target == nil {
		return fmt.Errorf("internal error: unknown section %q", section)
	}

	if !contains(*target, typeRef) {
		return fmt.Errorf("%s is not currently linked", typeRef)
	}

	*target = remove(*target, typeRef)

	if err := SaveProject(projectPath, config); err != nil {
		return err
	}

	return Sync(projectPath)
}

// Sync loads project.yaml and regenerates all tool configurations.
func Sync(projectPath string) error {
	config, err := LoadProject(projectPath)
	if err != nil {
		return err
	}

	installedPath, err := userdata.GetInstalledRoot()
	if err != nil {
		return fmt.Errorf("resolving installed root: %w", err)
	}

	tools := make([]integrations.ToolName, 0, len(config.Tools))
	for _, t := range config.Tools {
		tool, ok := integrations.ParseToolName(t)
		if !ok {
			return fmt.Errorf("unknown tool in project.yaml: %s", t)
		}
		tools = append(tools, tool)
	}

	// Convert ProjectConfig to a map for the Node generators.
	configMap := map[string]interface{}{
		"tools": config.Tools,
		"active": map[string]interface{}{
			"personas":  config.Active.Personas,
			"context":   config.Active.Context,
			"skills":    config.Active.Skills,
			"workflows": config.Active.Workflows,
			"prompts":   config.Active.Prompts,
		},
	}

	results, err := integrations.GenerateConfigs(tools, configMap, installedPath, projectPath)
	if err != nil {
		return fmt.Errorf("generating configs: %w", err)
	}

	// Print summary
	for _, r := range results {
		fmt.Printf("  %s: %d created, %d updated, %d symlinked\n",
			r.Tool, len(r.Created), len(r.Updated), len(r.Symlinked))
		for _, w := range r.Warnings {
			fmt.Printf("    warning: %s\n", w)
		}
	}

	return nil
}

// Status calls each tool's status check and returns the results.
func Status(projectPath string) ([]integrations.StatusResult, error) {
	config, err := LoadProject(projectPath)
	if err != nil {
		return nil, err
	}

	tools := make([]integrations.ToolName, 0, len(config.Tools))
	for _, t := range config.Tools {
		tool, ok := integrations.ParseToolName(t)
		if !ok {
			return nil, fmt.Errorf("unknown tool in project.yaml: %s", t)
		}
		tools = append(tools, tool)
	}

	return integrations.GetStatus(tools, projectPath)
}

File: src/core/linker.ts
import { join, dirname } from 'node:path';
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'node:fs';
import yaml from 'js-yaml';
import type { ToolName, GenerateResult, StatusResult } from '../types/integrations.js';
import { ALL_TOOLS } from '../types/integrations.js';

// ── Project config ──────────────────────────────────────────────────

export interface ActiveConfig {
  personas?: string[];
  context?: string[];
  skills?: string[];
  workflows?: string[];
  prompts?: string[];
}

export interface ProjectConfig {
  tools: string[];
  active: ActiveConfig;
}

const PROJECT_DIR = '.agentx';
const PROJECT_FILE = 'project.yaml';

export function projectConfigPath(projectPath: string): string {
  return join(projectPath, PROJECT_DIR, PROJECT_FILE);
}

export function loadProject(projectPath: string): ProjectConfig {
  const path = projectConfigPath(projectPath);
  const raw = readFileSync(path, 'utf-8');
  const data = yaml.load(raw) as ProjectConfig;
  return {
    tools: data.tools ?? [],
    active: {
      personas: data.active?.personas ?? [],
      context: data.active?.context ?? [],
      skills: data.active?.skills ?? [],
      workflows: data.active?.workflows ?? [],
      prompts: data.active?.prompts ?? [],
    },
  };
}

export function saveProject(
  projectPath: string,
  config: ProjectConfig,
): void {
  const path = projectConfigPath(projectPath);
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, yaml.dump(config, { lineWidth: -1 }), 'utf-8');
}

export function initProject(projectPath: string, tools: string[]): void {
  const agentxDir = join(projectPath, PROJECT_DIR);
  mkdirSync(agentxDir, { recursive: true });
  mkdirSync(join(agentxDir, 'overrides'), { recursive: true });

  const config: ProjectConfig = {
    tools,
    active: {
      personas: [],
      context: [],
      skills: [],
      workflows: [],
      prompts: [],
    },
  };
  saveProject(projectPath, config);
}

// ── Type management ─────────────────────────────────────────────────

function typeSection(typeRef: string): keyof ActiveConfig {
  const prefix = typeRef.split('/')[0];
  const map: Record<string, keyof ActiveConfig> = {
    personas: 'personas',
    context: 'context',
    skills: 'skills',
    workflows: 'workflows',
    prompts: 'prompts',
  };
  const section = map[prefix];
  if (!section) {
    throw new Error(`Unknown type prefix: "${prefix}". Expected: personas, context, skills, workflows, or prompts.`);
  }
  return section;
}

export async function addType(projectPath: string, typeRef: string): Promise<void> {
  const config = loadProject(projectPath);
  const section = typeSection(typeRef);
  const list = config.active[section] ?? [];
  if (list.includes(typeRef)) {
    throw new Error(`Type "${typeRef}" is already linked.`);
  }
  list.push(typeRef);
  config.active[section] = list;
  saveProject(projectPath, config);
  await sync(projectPath);
}

export async function removeType(projectPath: string, typeRef: string): Promise<void> {
  const config = loadProject(projectPath);
  const section = typeSection(typeRef);
  const list = config.active[section] ?? [];
  if (!list.includes(typeRef)) {
    throw new Error(`Type "${typeRef}" is not linked.`);
  }
  config.active[section] = list.filter((t) => t !== typeRef);
  saveProject(projectPath, config);
  await sync(projectPath);
}

// ── Sync & Status ───────────────────────────────────────────────────

export async function sync(projectPath: string): Promise<GenerateResult[]> {
  const config = loadProject(projectPath);
  const { getInstalledRoot } = await import('./userdata.js');
  const installedPath = getInstalledRoot();

  const { generate } = await import('../integrations/index.js');
  const results: GenerateResult[] = [];

  for (const toolName of config.tools) {
    try {
      const result = await generate({
        toolName,
        projectConfig: config,
        installedPath,
        projectPath,
      });
      results.push(result as GenerateResult);
    } catch (err) {
      results.push({
        tool: toolName as ToolName,
        created: [],
        updated: [],
        symlinked: [],
        warnings: [String(err)],
      });
    }
  }
  return results;
}

export async function status(projectPath: string): Promise<StatusResult[]> {
  const config = loadProject(projectPath);

  const { status: getStatus } = await import('../integrations/index.js');
  const results: StatusResult[] = [];

  for (const toolName of config.tools) {
    try {
      const result = await getStatus({ toolName, projectPath });
      results.push(result as StatusResult);
    } catch (err) {
      results.push({
        tool: toolName,
        status: 'error',
        files: [],
        symlinks: { total: 0, valid: 0 },
      });
    }
  }
  return results;
}
